docs: add agentgateway example (#376)
kagent-dev/website: src/app/docs/kagent/examples/agentgateway/page.mdx

---
title: "agentgateway"
pageOrder: 1
description: "Add governance to your kagent deployment with agentgateway"
---

import { Tabs, Tab } from '@/components/mdx/tabs';

export const metadata = {
  title: "Using agentgateway with kagent",
  description: "Learn how to add governance to your kagent deployment using agentgateway",
  author: "kagent.dev"
};

# Using agentgateway with kagent

> As your kagent deployment grows, you may need governance over how your agents communicate with LLM providers, things like access control, rate limiting, audit logging, and observability. [Agentgateway](https://agentgateway.dev) is a proxy purpose-built for AI workloads that sits between your kagent agents and your LLM provider to provide exactly this.

Reviewer: ```suggestion
As your kagent deployment grows, you might need governance over how your agents communicate with LLM providers. [Agentgateway](https://agentgateway.dev) is a proxy purpose-built for AI workloads that sits between your kagent agents and your LLM provider. This way, you can apply AgentgatewayPolicy for things like access control, rate limiting, audit logging, and observability. 
```

## Prerequisites

1. A running kagent installation. If you haven't installed kagent yet, follow the [quick start](/docs/kagent/getting-started/quickstart) guide first.
2. Follow the [agentgateway installation guide](https://agentgateway.dev/docs/kubernetes/latest/quickstart/install/) to install agentgateway in your cluster.

> 3. Follow the [Ollama guide](https://agentgateway.dev/docs/kubernetes/latest/llm/providers/ollama/) to install and configure Ollama as your LLM provider.

Reviewer: Maybe say that it could be any LLM provider, this guide just uses Ollama as an example.
```suggestion
3. Set up an LLM provider with agentgateway. This guide uses [the Ollama setup](https://agentgateway.dev/docs/kubernetes/latest/llm/providers/ollama/) as an example.
```

## Architecture

> Once set up, kagent agent pods route all LLM requests through agentgateway (running in the `agentgateway-system` namespace). Agentgateway enforces your policies, auth/authz, rate limiting, audit logging, and observability. Before forwarding requests to Ollama on the host.

Reviewer: ```suggestion
After setup, agentgateway handles the traffic between your kagent agents and your LLM providers. Agentgateway enforces your policies, auth/authz, rate limiting, audit logging, and observability.
```

Reviewer: maybe a simple mermaid.js diagram?

Author: sure [903597a](https://github.com/kagent-dev/website/pull/376/commits/903597a8ceb89ef1577a58c240f7a8ebb3cd04d7)

## Configure kagent to use agentgateway

With agentgateway installed, point kagent at the agentgateway proxy instead of directly at Ollama.

1. If you installed kagent without agentgateway, upgrade your installation to route through the proxy.

   ```shell

>    helm upgrade kagent oci://ghcr.io/kagent-dev/kagent/helm/kagent \

Reviewer: might need the `--reuse-values` flag so that the set commands don't overwrite any other sets.

     --namespace kagent \
     --set providers.default=ollama \
     --set providers.ollama.baseUrl=http://agentgateway-proxy.agentgateway-system.svc.cluster.local/v1 \
     --set providers.ollama.apiKey=dummy
   ```

2. Create a `ModelConfig` that points to Ollama via the agentgateway proxy.

   ```yaml
   kubectl apply -f- <<EOF
   apiVersion: kagent.dev/v1alpha2
   kind: ModelConfig
   metadata:
     name: llama3-model-config
     namespace: kagent
   spec:
     model: llama3
     provider: Ollama
     ollama:
       host: agentgateway-proxy.agentgateway-system.svc.cluster.local
   EOF
   ```

3. Verify that kagent is still accessible and correctly functioning.

   <Tabs tabs={[
     { id: 'loadbalancer', label: 'Cloud Provider LoadBalancer' },
     { id: 'port-forward', label: 'Port-forward for local testing' }
   ]} />

   <div id="loadbalancer-tab">
   ```shell
   export INGRESS_GW_ADDRESS=$(kubectl get svc -n kagent kagent-ui -o jsonpath="{.spec.clusterIP}")
   echo $INGRESS_GW_ADDRESS
   ```
   </div>

   <div id="port-forward-tab">
   ```shell
   kubectl port-forward -n kagent service/kagent-ui 8082:8080
   ```
   </div>

> 4. Open the kagent UI and try the default `k8s-agent` to confirm requests are flowing through agentgateway end-to-end.

Reviewer: ```suggestion
4. [Open the kagent UI](/docs/kagent/observability/launch-ui).
5. Start a chat with an agent such as `k8s-agent` to confirm that requests flow through agentgateway.
```

   ```yaml
   kubectl apply -f - <<EOF
   apiVersion: agentgateway.dev/v1alpha1
   kind: AgentgatewayPolicy
   metadata:
     name: prompt-guard
     namespace: agentgateway-system
   spec:
     targetRefs:
     - group: gateway.networking.k8s.io
       kind: HTTPRoute
       name: ollama
     backend:
       ai:
         promptGuard:
           request:
           - response:
               message: "Rejected due to inappropriate content"
             regex:
               action: Reject
               matches:
               - "email"
   EOF
   ```

2. Verify the policy by sending a prompt to your agent through the kagent UI that includes the word `email`. You get a `403` response.

   ![kagent rejected content](/images/kagent-rejected-content.png "kagent rejected content")

## Cleanup

To remove agentgateway-related resources from your kagent setup:

```shell
kubectl delete agentgatewaypolicy prompt-guard -n agentgateway-system
kubectl delete modelconfig llama3-model-config -n kagent
```

> To also remove agentgateway itself, follow the [agentgateway uninstall guide](https://agentgateway.dev/docs/kubernetes/latest/operations/uninstall/).

Reviewer: Step format, and then also need to unset agentgateway as the proxy between the LLM provider.

1. Remove agentgateway-related resources from your kagent setup.

   ```shell
   kubectl delete agentgatewaypolicy prompt-guard -n agentgateway-system
   kubectl delete modelconfig llama3-model-config -n kagent
   ```

2. If you no longer need agentgateway, [uninstall agentgateway](https://agentgateway.dev/docs/kubernetes/latest/operations/uninstall/).

3. Upgrade your kagent installation to unset agentgateway as the proxy in front of your LLM provider.

<Mermaid chart={`
flowchart LR
 subgraph KindCluster["kind cluster"]
        kagentPods["kagent agent pods"]
        agentGateway["agentgateway<br>(agentgateway-system ns)<br>• auth / authz<br>• rate limiting<br>• audit logging<br>• observability"]
  end
    kagentPods --> agentGateway
    agentGateway --> ollama["Ollama<br>(host)"]
     kagentPods:::internal
     agentGateway:::internal
     ollama:::external
    classDef cluster stroke:#818cf8,fill:#eef2ff
    classDef internal stroke:#a78bfa,fill:#f5f3ff
    classDef external stroke:#fb923c,fill:#fff7ed
    style kagentPods stroke:#a78bfa,fill:#ffffff
    style agentGateway fill:#ffffff,stroke:#AA00FF
    style ollama fill:#ffffff,stroke:#00C853
    style KindCluster stroke:#2962FF,fill:#ffffff
`} />

## Configure kagent to use agentgateway

With agentgateway installed, point kagent at the agentgateway proxy instead of directly at Ollama.

1. If you installed kagent without agentgateway, upgrade your installation to route through the proxy.

   ```shell

>    helm upgrade kagent oci://ghcr.io/kagent-dev/kagent/helm/kagent \

Reviewer 2: You shouldn't need to do this step if you manually define the ModelConfig with the host!

>      --reuse-values

Reviewer 2: ```suggestion
     --reuse-values \
```

     --set providers.default=ollama \
     --set providers.ollama.baseUrl=http://agentgateway-proxy.agentgateway-system.svc.cluster.local/v1 \
     --set providers.ollama.apiKey=dummy
   ```

2. Create a `ModelConfig` that points to Ollama via the agentgateway proxy.

   ```yaml
   kubectl apply -f- <<EOF
   apiVersion: kagent.dev/v1alpha2
   kind: ModelConfig
   metadata:
     name: llama3-model-config
     namespace: kagent
   spec:
     model: llama3
     provider: Ollama
     ollama:
       host: agentgateway-proxy.agentgateway-system.svc.cluster.local
   EOF
   ```

3. Verify that kagent is still accessible and correctly functioning.

   <Tabs tabs={[
     { id: 'loadbalancer', label: 'Cloud Provider LoadBalancer' },
     { id: 'port-forward', label: 'Port-forward for local testing' }
   ]} />

   <div id="loadbalancer-tab">
   ```shell
   export INGRESS_GW_ADDRESS=$(kubectl get svc -n kagent kagent-ui -o jsonpath="{.spec.clusterIP}")
   echo $INGRESS_GW_ADDRESS
   ```
   </div>

   <div id="port-forward-tab">
   ```shell
   kubectl port-forward -n kagent service/kagent-ui 8082:8080
   ```
   </div>

4. [Open the kagent UI](/docs/kagent/observability/launch-ui).

> 5. Start a chat with an agent such as `k8s-agent` to confirm that requests flow through agentgateway.

Reviewer 2: So the helm upgrade step will change all the agents to go through agentgateway. It might be worth showing how you can make a specific agent use the ModelConfig by editing the agent/creating a new agent instead. The basic getting started guide uses the cli to install https://kagent.dev/docs/kagent/getting-started/quickstart (so the helm upgrade might not behave as expected)

   ![kagent default k8s-agent UI](/images/kagent-default-k8s-agent.png "kagent default k8s-agent UI")

## Apply governance policies

With agentgateway in place, you can now apply policies to govern how your kagent agents interact with your LLM provider.

### Block requests with PII

1. Create an `AgentgatewayPolicy` resource to reject any request that contains PII, such as an email address. For more policy examples, see the [agentgateway guardrails docs](https://agentgateway.dev/docs/kubernetes/latest/llm/guardrails/regex/#block-requests-with-pii).

   ```yaml
   kubectl apply -f - <<EOF
   apiVersion: agentgateway.dev/v1alpha1
   kind: AgentgatewayPolicy
   metadata:
     name: prompt-guard
     namespace: agentgateway-system
   spec:
     targetRefs:
     - group: gateway.networking.k8s.io
       kind: HTTPRoute
       name: ollama
     backend:
       ai:
         promptGuard:
           request:
           - response:
               message: "Rejected due to inappropriate content"
             regex:

>                action: Reject

Reviewer 2: Let's switch this to mask and then show the chat will have `EMAIL` masked out!
